util/codec/codec.go

```diff
@@ -352,6 +352,240 @@ func encodeHashChunkRowIdx(sc *stmtctx.StatementContext, row chunk.Row, tp *type
 	return
 }
 
+func writeColumn(w io.Writer, flag, data []byte) (err error) {
+	_, err = w.Write(flag)
+	if err != nil {
+		return
+	}
+	_, err = w.Write(data)
+	return
+}
+
+// HashChunkColumns writes the encoded value of each row's column, which of index `colIdx`, to w.
+// `nullBitmap` used to save the rows whose column of index `colIdx` is `Null`.
+func HashChunkColumns(sc *stmtctx.StatementContext, w []io.Writer, chk *chunk.Chunk, tp *types.FieldType, colIdx int, buf, nullBitmap []byte) (err error) {
+	var (
+		b    []byte
+		rows int
+		row  chunk.Row
+	)
+	rows = chk.NumRows()
+	switch tp.Tp {
+	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeYear:
+		for i := 0; i < rows; i++ {
+			row = chk.GetRow(i)
+			if row.IsNull(colIdx) {
+				buf[0], b = NilFlag, nil
+				nullBitmap[i/8] |= 1 << uint(i%8)
+			} else {
+				buf[0] = varintFlag
+				if mysql.HasUnsignedFlag(tp.Flag) {
+					if integer := row.GetInt64(colIdx); integer < 0 {
+						buf[0] = uvarintFlag
+					}
+				}
+				b = row.GetRaw(colIdx)
+			}
+
+			err = writeColumn(w[i], buf, b)
```

Comment:
How about removing this `writeColumn` method and creating our own hash methods?
We can use `b` to compute the hash sum here directly instead of copying it into `w[i]` and computing it later.

Reply:
Yes, this may be a good suggestion. According to my `cpuprofile` result, `hash.Hash64.Write` consumes most of the CPU time of `hashRowContainer.PutChunk`.
I will try it and give a benchmark result later.

Reply:
Emm, you can ignore this comment since it's my mistake.
After investigation, I found there is no data copying since `[fnv.Write](https://golang.org/src/hash/fnv/fnv.go#L99)` works in a streaming manner.